File: backend/src/main/java/com/technicalescaperoom/backend/service/TeamService.java
package com.technicalescaperoom.backend.service;

import com.technicalescaperoom.backend.dto.admin.*;
import com.technicalescaperoom.backend.entity.Event;
import com.technicalescaperoom.backend.entity.Player;
import com.technicalescaperoom.backend.entity.Team;
import com.technicalescaperoom.backend.enums.GameEventType;
import com.technicalescaperoom.backend.enums.PlayerStatus;
import com.technicalescaperoom.backend.enums.TeamStatus;
import com.technicalescaperoom.backend.exception.ResourceNotFoundException;
import com.technicalescaperoom.backend.repository.EventRepository;
import com.technicalescaperoom.backend.repository.PlayerRepository;
import com.technicalescaperoom.backend.repository.TeamRepository;
import com.technicalescaperoom.backend.util.TeamCodeGenerator;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class TeamService {

    private final EventRepository eventRepository;
    private final TeamRepository teamRepository;
    private final PlayerRepository playerRepository;
    private final TeamCodeGenerator teamCodeGenerator;
    private final AuditService auditService;
    private final GameStateService gameStateService;

    @Transactional
    public TeamDetailResponse createTeam(Long eventId, CreateTeamRequest request) {
        Event event = eventRepository.findById(eventId)
                .orElseThrow(() -> new ResourceNotFoundException("Event not found with ID: " + eventId));

        if (request.getPlayer1DisplayName() == null || request.getPlayer1DisplayName().isBlank()) {
            throw new IllegalArgumentException("Player 1 display name is required");
        }
        if (request.getPlayer2DisplayName() == null || request.getPlayer2DisplayName().isBlank()) {
            throw new IllegalArgumentException("Player 2 display name is required");
        }

        String teamCode;
        if (request.getCustomTeamCode() != null && !request.getCustomTeamCode().isBlank()) {
            teamCode = request.getCustomTeamCode().trim().toUpperCase();
            if (teamRepository.existsByEventIdAndTeamCode(eventId, teamCode)) {
                throw new IllegalArgumentException("Team code '" + teamCode + "' already exists for this event");
            }
        } else {
            long currentCount = teamRepository.countByEventId(eventId);
            teamCode = teamCodeGenerator.generateUniqueTeamCode(currentCount,
                    code -> teamRepository.existsByEventIdAndTeamCode(eventId, code));
        }

        Team team = Team.builder()
                .event(event)
                .teamCode(teamCode)
                .teamName(request.getTeamName().trim())
                .status(TeamStatus.REGISTERED)
                .build();

        Team savedTeam = teamRepository.save(team);

        Player p1 = Player.builder()
                .team(savedTeam)
                .playerNumber(1)
                .displayName(request.getPlayer1DisplayName().trim())
                .status(PlayerStatus.INACTIVE)
                .build();

        Player p2 = Player.builder()
                .team(savedTeam)
                .playerNumber(2)
                .displayName(request.getPlayer2DisplayName().trim())
                .status(PlayerStatus.INACTIVE)
                .build();

        playerRepository.save(p1);
        playerRepository.save(p2);

        gameStateService.initializeTeamGameState(savedTeam);

        auditService.logEvent(GameEventType.TEAM_CREATED, event, savedTeam, null,
                "{\"teamCode\":\"" + teamCode + "\",\"teamName\":\"" + savedTeam.getTeamName() + "\"}");

        return mapToDetailResponse(savedTeam);
    }

    @Transactional(readOnly = true)
    public TeamDetailResponse getTeamById(Long teamId) {
        Team team = findTeamOrThrow(teamId);
        return mapToDetailResponse(team);
    }

    @Transactional(readOnly = true)
    public List<TeamResponse> getTeamsByEventId(Long eventId) {
        if (!eventRepository.existsById(eventId)) {
            throw new ResourceNotFoundException("Event not found with ID: " + eventId);
        }

        return teamRepository.findByEventId(eventId).stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public TeamDetailResponse updateTeam(Long teamId, UpdateTeamRequest request) {
        Team team = findTeamOrThrow(teamId);

        team.setTeamName(request.getTeamName().trim());
        team.setStatus(request.getStatus());

        List<Player> players = playerRepository.findByTeamId(teamId);
        for (Player p : players) {
            if (p.getPlayerNumber() == 1 && request.getPlayer1DisplayName() != null) {
                p.setDisplayName(request.getPlayer1DisplayName().trim());
                playerRepository.save(p);
            } else if (p.getPlayerNumber() == 2 && request.getPlayer2DisplayName() != null) {
                p.setDisplayName(request.getPlayer2DisplayName().trim());
                playerRepository.save(p);
            }
        }

        Team updatedTeam = teamRepository.save(team);

        auditService.logEvent(GameEventType.TEAM_UPDATED, updatedTeam.getEvent(), updatedTeam, null,
                "{\"status\":\"" + updatedTeam.getStatus() + "\"}");

        return mapToDetailResponse(updatedTeam);
    }

    @Transactional
    public TeamDetailResponse updateTeamStatus(Long teamId, TeamStatus status) {
        Team team = findTeamOrThrow(teamId);
        team.setStatus(status);

        Team updatedTeam = teamRepository.save(team);

        GameEventType auditType = (status == TeamStatus.ACTIVE)
                ? GameEventType.TEAM_ACTIVATED
                : GameEventType.TEAM_STATUS_CHANGED;

        auditService.logEvent(auditType, updatedTeam.getEvent(), updatedTeam, null,
                "{\"newStatus\":\"" + status + "\"}");

        return mapToDetailResponse(updatedTeam);
    }

    @Transactional
    public void deleteTeam(Long teamId) {
        Team team = findTeamOrThrow(teamId);
        auditService.logEvent(GameEventType.TEAM_DELETED, team.getEvent(), team, null,
                "{\"deletedTeamCode\":\"" + team.getTeamCode() + "\"}");

        playerRepository.deleteByTeamId(teamId);
        teamRepository.delete(team);
    }

    public Team findTeamOrThrow(Long teamId) {
        return teamRepository.findById(teamId)
                .orElseThrow(() -> new ResourceNotFoundException("Team not found with ID: " + teamId));
    }

    private TeamResponse mapToResponse(Team team) {
        List<Player> players = playerRepository.findByTeamId(team.getId());
        String p1Name = players.stream()
                .filter(p -> p.getPlayerNumber() == 1)
                .map(Player::getDisplayName)
                .findFirst().orElse("");
        String p2Name = players.stream()
                .filter(p -> p.getPlayerNumber() == 2)
                .map(Player::getDisplayName)
                .findFirst().orElse("");

        return TeamResponse.builder()
                .id(team.getId())
                .eventId(team.getEvent().getId())
                .teamCode(team.getTeamCode())
                .teamName(team.getTeamName())
                .status(team.getStatus())
                .player1DisplayName(p1Name)
                .player2DisplayName(p2Name)
                .createdAt(team.getCreatedAt())
                .updatedAt(team.getUpdatedAt())
                .build();
    }

    private TeamDetailResponse mapToDetailResponse(Team team) {
        List<Player> players = playerRepository.findByTeamId(team.getId());
        List<PlayerDto> playerDtos = players.stream()
                .map(p -> PlayerDto.builder()
                        .id(p.getId())
                        .playerNumber(p.getPlayerNumber())
                        .displayName(p.getDisplayName())
                        .status(p.getStatus())
                        .createdAt(p.getCreatedAt())
                        .updatedAt(p.getUpdatedAt())
                        .build())
                .collect(Collectors.toList());

        return TeamDetailResponse.builder()
                .id(team.getId())
                .eventId(team.getEvent().getId())
                .eventName(team.getEvent().getName())
                .teamCode(team.getTeamCode())
                .teamName(team.getTeamName())
                .status(team.getStatus())
                .completedAt(team.getCompletedAt())
                .players(playerDtos)
                .createdAt(team.getCreatedAt())
                .updatedAt(team.getUpdatedAt())
                .build();
    }
}

File: backend/services/teamService.js
const eventRepository = require('../repositories/eventRepository');
const teamRepository = require('../repositories/teamRepository');
const playerRepository = require('../repositories/playerRepository');
const gameStateService = require('./gameStateService');
const auditService = require('./auditService');
const { ResourceNotFoundException } = require('../middleware/errorHandler');

class TeamService {
  async createTeam(eventId, request) {
    const event = await eventRepository.findById(eventId);
    if (!event) {
      throw new ResourceNotFoundException(`Event not found with ID: ${eventId}`);
    }

    if (!request.player1DisplayName || !request.player1DisplayName.trim()) {
      const err = new Error('Player 1 display name is required');
      err.status = 400;
      throw err;
    }
    if (!request.player2DisplayName || !request.player2DisplayName.trim()) {
      const err = new Error('Player 2 display name is required');
      err.status = 400;
      throw err;
    }

    let teamCode;
    if (request.customTeamCode && request.customTeamCode.trim()) {
      teamCode = request.customTeamCode.trim().toUpperCase();
      const existing = await teamRepository.findByEventIdAndTeamCode(eventId, teamCode);
      if (existing) {
        const err = new Error(`Team code '${teamCode}' already exists for this event`);
        err.status = 400;
        throw err;
      }
    } else {
      const currentCount = await teamRepository.countByEventId(eventId);
      teamCode = await this._generateUniqueTeamCode(eventId, currentCount);
    }

    const team = await teamRepository.createTeam({
      eventId: event.id,
      teamCode,
      teamName: (request.teamName || teamCode).trim(),
      status: 'REGISTERED',
      gameState: 'NOT_STARTED'
    });

    const p1 = await playerRepository.create({
      teamId: team.id,
      playerNumber: 1,
      displayName: request.player1DisplayName.trim(),
      status: 'INACTIVE'
    });

    const p2 = await playerRepository.create({
      teamId: team.id,
      playerNumber: 2,
      displayName: request.player2DisplayName.trim(),
      status: 'INACTIVE'
    });

    const fullTeam = await teamRepository.findById(team.id);
    await gameStateService.initializeTeamGameState(fullTeam);

    await auditService.logEvent(
      'TEAM_CREATED',
      event,
      fullTeam,
      null,
      JSON.stringify({ teamCode, teamName: fullTeam.teamName }),
      'ADMIN'
    );

    return await this._mapToDetailResponse(fullTeam);
  }

  async getTeamById(teamId) {
    const team = await this.findTeamOrThrow(teamId);
    return await this._mapToDetailResponse(team);
  }

  async getTeamsByEventId(eventId) {
    const event = await eventRepository.findById(eventId);
    if (!event) {
      throw new ResourceNotFoundException(`Event not found with ID: ${eventId}`);
    }

    const teams = await teamRepository.findByEventId(eventId);
    const result = [];
    for (const team of teams) {
      result.push(await this._mapToResponse(team));
    }
    return result;
  }

  async updateTeam(teamId, request) {
    const team = await this.findTeamOrThrow(teamId);

    if (request.teamName) team.teamName = request.teamName.trim();
    if (request.status) team.status = request.status;

    const players = await playerRepository.findByTeamId(teamId);
    for (const p of players) {
      if (p.playerNumber === 1 && request.player1DisplayName) {
        p.displayName = request.player1DisplayName.trim();
        await playerRepository.save(p);
      } else if (p.playerNumber === 2 && request.player2DisplayName) {
        p.displayName = request.player2DisplayName.trim();
        await playerRepository.save(p);
      }
    }

    const updatedTeam = await teamRepository.save(team);

    await auditService.logEvent(
      'TEAM_UPDATED',
      updatedTeam.event,
      updatedTeam,
      null,
      JSON.stringify({ status: updatedTeam.status }),
      'ADMIN'
    );

    return await this._mapToDetailResponse(updatedTeam);
  }

  async updateTeamStatus(teamId, status) {
    const team = await this.findTeamOrThrow(teamId);
    team.status = status;

    const updatedTeam = await teamRepository.save(team);

    const auditType = (status === 'ACTIVE') ? 'TEAM_ACTIVATED' : 'TEAM_STATUS_CHANGED';
    await auditService.logEvent(
      auditType,
      updatedTeam.event,
      updatedTeam,
      null,
      JSON.stringify({ newStatus: status }),
      'ADMIN'
    );

    return await this._mapToDetailResponse(updatedTeam);
  }

  async deleteTeam(teamId) {
    const team = await this.findTeamOrThrow(teamId);

    await auditService.logEvent(
      'TEAM_DELETED',
      team.event,
      team,
      null,
      JSON.stringify({ deletedTeamCode: team.teamCode }),
      'ADMIN'
    );

    await playerRepository.deleteByTeamId(teamId);
    await teamRepository.deleteById(teamId);
  }

  async findTeamOrThrow(teamId) {
    const team = await teamRepository.findById(teamId);
    if (!team) {
      throw new ResourceNotFoundException(`Team not found with ID: ${teamId}`);
    }
    return team;
  }

  async _generateUniqueTeamCode(eventId, currentCount) {
    let index = currentCount + 1;
    while (true) {
      const code = `TEAM-${String(index).padStart(3, '0')}`;
      const existing = await teamRepository.findByEventIdAndTeamCode(eventId, code);
      if (!existing) {
        return code;
      }
      index++;
    }
  }

  async _mapToResponse(team) {
    const players = await playerRepository.findByTeamId(team.id);
    const p1 = players.find(p => p.playerNumber === 1);
    const p2 = players.find(p => p.playerNumber === 2);

    return {
      id: team.id,
      eventId: team.eventId,
      teamCode: team.teamCode,
      teamName: team.teamName,
      status: team.status,
      player1DisplayName: p1 ? p1.displayName : '',
      player2DisplayName: p2 ? p2.displayName : '',
      createdAt: team.createdAt,
      updatedAt: team.updatedAt
    };
  }

  async _mapToDetailResponse(team) {
    const players = await playerRepository.findByTeamId(team.id);
    const playerDtos = players.map(p => ({
      id: p.id,
      playerNumber: p.playerNumber,
      displayName: p.displayName,
      status: p.status,
      createdAt: p.createdAt,
      updatedAt: p.updatedAt
    }));

    return {
      id: team.id,
      eventId: team.eventId,
      eventName: team.event ? team.event.name : null,
      teamCode: team.teamCode,
      teamName: team.teamName,
      status: team.status,
      completedAt: team.completedAt,
      players: playerDtos,
      createdAt: team.createdAt,
      updatedAt: team.updatedAt
    };
  }
}

module.exports = new TeamService();
